[OSDOCS-6942]:OSD on GCP can be installed into Shared VPC (XPN)

mletalie opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OCPBUGS#<jira-issue-id>][OSDOCS#<jira-issue-id>]: <short-description-of-the-pr> --->

<!--- If your changes apply to the latest release and/or in-development version of OpenShift, open your PR against the `main` branch.

* For more details about the information requested in this template, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/create_or_edit_content.adoc#submit-PR --->

Version(s):
<!--- Specify the version or versions of OpenShift your PR applies to. -->
4.14+
Issue:
<!--- Add a link to the Bugzilla, Jira, or GitHub issue, if applicable. --->
https://issues.redhat.com/browse/OSDOCS-6942
Link to docs preview:
<!--- Add direct link(s) to the exact page(s) with updated content from the preview build. --->
https://66312--docspreview.netlify.app/openshift-dedicated/latest/osd_install_access_delete_cluster/creating-a-gcp-cluster#osd-create-gcp-cluster-ccs_osd-creating-a-cluster-on-gcp
**(Scroll to step 15. Changes happen between steps 15-17, and a note is added after step 20).** 

https://66312--docspreview.netlify.app/openshift-dedicated/latest/osd_install_access_delete_cluster/osd-deleting-a-cluster
**Note at end of steps.**
QE review:
- [x] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

Additional information:
<!--- Optional: Include additional context or expand the description here.--->
![image](https://github.com/openshift/openshift-docs/assets/122639474/57dc4565-3eef-40c8-afb7-c7da14494682)

<!--- After you open your PR, ask for review from the OpenShift docs team:
  For community authors: Tag @openshift/team-documentation in a GitHub comment.--->

mletalie commented:
Hello [@svmrh](https://github.com/svmrh),
     Please see the draft for this topic here (starting on step 15) https://66312--docspreview.netlify.app/openshift-dedicated/latest/osd_install_access_delete_cluster/creating-a-gcp-cluster#osd-create-gcp-cluster-ccs_osd-creating-a-cluster-on-gcp

mletalie commented:
> @mletalie : Thanks for putting this together.
> 
> 1. The placement of the first "IMPORTANT" note if fine. Yes, it will be good to point users to the GCP documentation on Shared VPC. Here's the link to GCP documentation explaining the steps to enable a host project.
>    https://cloud.google.com/vpc/docs/provisioning-shared-vpc#set-up-shared-vpc
> 
> Do you think it makes sense to also add a link to general overview on what is Shared VPC in Google Cloud? https://cloud.google.com/vpc/docs/shared-vpc
> 
> 2. For the IMPORTANT note in 15.b, can we make it explicit that the cluster admin has to complete the steps within the cluster configuration wizard and hit submit. I have made minor edits. Please check if it looks okay
> 
> > Once you complete the steps within the cluster configuration wizard **and click "Create Cluster", the cluster will go into "Installation Pending" state**. At this point, you must contact the VPC administrator of the host project, who must grant the **dynamically generated** service account the following permissions: Computer Network Administrator, Compute Security Administrator, and DNS Administrator. The administrator has 30 days to grant the listed permissions before the cluster creation fails. For information about shared GPC VPC permissions, see Provision shared VPC.
> 
> 3. the placement of the second IMPORTANT note is fine here (i.e. in step 15.b). Placing it towards the end of all the steps (i.e. after step 20) might be a little off context. What are your thoughts?
> 4. Yeah, no need of adding references to the Shared VPC option in step 16. This is because the users can install into an existing VPC (i.e. do step 16) and still skip the Shared VPC configuration.
> 
> Let's share this MR in tomorrow's (Thursday, Oct 19th) team meeting and request everyone to review changes.

Thanks @svmrh, will take a look at your responses and get back to you ASAP. Appreciate it!

mletalie commented:
Hello  @svmrh,
    Thanks again for feedback. See responses below: 


> > @mletalie : Thanks for putting this together.
> > 
> > 1. The placement of the first "IMPORTANT" note if fine. Yes, it will be good to point users to the GCP documentation on Shared VPC. Here's the link to GCP documentation explaining the steps to enable a host project.
> >    https://cloud.google.com/vpc/docs/provisioning-shared-vpc#set-up-shared-vpc
> > **DONE**
> > Do you think it makes sense to also add a link to general overview on what is Shared VPC in Google Cloud? https://cloud.google.com/vpc/docs/shared-vpc
> > 
> > 2. For the IMPORTANT note in 15.b, can we make it explicit that the cluster admin has to complete the steps within the cluster configuration wizard and hit submit. I have made minor edits. Please check if it looks okay
> > 
> > > Once you complete the steps within the cluster configuration wizard **and click "Create Cluster", the cluster will go into "Installation Pending" state**. At this point, you must contact the VPC administrator of the host project, who must grant the **dynamically generated** service account the following permissions: Computer Network Administrator, Compute Security Administrator, and DNS Administrator. The administrator has 30 days to grant the listed permissions before the cluster creation fails. For information about shared GPC VPC permissions, see Provision shared VPC.
> > **DONE**
> > 
> > 
> > 3. the placement of the second IMPORTANT note is fine here (i.e. in step 15.b). Placing it towards the end of all the steps (i.e. after step 20) might be a little off context. What are your thoughts?
**I think it is good as is. Thanks!** 
> > 4. Yeah, no need of adding references to the Shared VPC option in step 16. This is because the users can install into an existing VPC (i.e. do step 16) and still skip the Shared VPC configuration.
> > Sounds good. Appreciate the clarity! 
> > Let's share this MR in tomorrow's (Thursday, Oct 19th) team meeting and request everyone to review changes.
> 
> Thanks @svmrh, will take a look at your responses and get back to you ASAP. Appreciate it!

**Thanks @svmrh, please see comments above. Appreciate it!**

mletalie commented:
Hello reviewers, please see https://redhat-internal.slack.com/archives/D04KG1N67MZ/p1698090978994849 which describes why I needed to conditionalize out the hyperlink to keep the build from breaking.

mletalie commented:
Hello @jianli-wei, when you get a moment can you please review the doc changes brought forth in these docs? Thanks!

svmrh commented:
LGTM.

A few minor comments -
1. Does it makes sense to repeat this note on [Deleting an OpenShift Dedicated cluster](https://66312--docspreview.netlify.app/openshift-dedicated/latest/osd_install_access_delete_cluster/osd-deleting-a-cluster) page as well?
![image](https://github.com/openshift/openshift-docs/assets/130616558/67570ae6-8c3c-4576-a6b3-408932ac7bb5)

2. @ckandag - Should we tweak step #16 a little to say something like "If you are installing into a GCP shared VPC, the VPC name and subnets are being shared from the host project..."
![image](https://github.com/openshift/openshift-docs/assets/130616558/47f5f37f-0e65-47a3-83b3-bf011b57e795)

@jianli-wei - can you please review the doc changes and give a QE :thumbsup:

Thanks @mletalie for driving the documentation changes for this feature.

jianli-wei commented on modules/osd-create-cluster-ccs.adoc:
```diff
@@ -284,6 +306,13 @@ In the event of critical security concerns that significantly impact the securit
+ifdef::osd-on-gcp[]
+[NOTE]
+====
+If you delete a cluster that was installed into a GCP shared VPC, you must inform the VPC owner of the host project to remove the service account and the roles granted to the service account that was referenced during the creation of the cluster.
```
In fact, the service account itself would be deleted during the cluster deletion. So suggest to put it like below, 

... you must inform the VPC owner of the host project to remove the IAM policy binding of the service account and its binding roles in the host project.

mletalie replied:
Hello @svmrh,
For point #1, I have added a note to that doc: 
![image](https://github.com/openshift/openshift-docs/assets/122639474/e7c57191-4fb6-45db-965d-a73739f86747)

mletalie replied:
For point #2, I have added a note: 
![image](https://github.com/openshift/openshift-docs/assets/122639474/97ce29c4-cf71-4bd2-b66c-edb02320bb55)

mletalie replied:
> In fact, the service account itself would be deleted during the cluster deletion. So suggest to put it like below,
> 
> ... you must inform the VPC owner of the host project to remove the IAM policy binding of the service account and its binding roles in the host project.

Thanks @jianli-wei,
   I applied your suggestion with just a slight tweak in wording to fit the narrative prior to that note. Thanks for the clarity.   
![image](https://github.com/openshift/openshift-docs/assets/122639474/bb4a7c3f-64f5-46f3-a5b5-90eb2d38c536)

mletalie commented:
> LGTM.
> 
> A few minor comments -
> 
> 1. Does it makes sense to repeat this note on [Deleting an OpenShift Dedicated cluster](https://66312--docspreview.netlify.app/openshift-dedicated/latest/osd_install_access_delete_cluster/osd-deleting-a-cluster) page as well?
>    ![image](https://user-images.githubusercontent.com/130616558/279811202-67570ae6-8c3c-4576-a6b3-408932ac7bb5.png)
> 2. @ckandag - Should we tweak step 16 a little to say something like "If you are installing into a GCP shared VPC, the VPC name and subnets are being shared from the host project..."
>    ![image](https://user-images.githubusercontent.com/130616558/279811446-47f5f37f-0e65-47a3-83b3-bf011b57e795.png)
> 
> @jianli-wei - can you please review the doc changes and give a QE 👍
> 
> Thanks @mletalie for driving the documentation changes for this feature.



> LGTM.
> 
> A few minor comments -
> 
> 1. Does it makes sense to repeat this note on [Deleting an OpenShift Dedicated cluster](https://66312--docspreview.netlify.app/openshift-dedicated/latest/osd_install_access_delete_cluster/osd-deleting-a-cluster) page as well?
>    ![image](https://user-images.githubusercontent.com/130616558/279811202-67570ae6-8c3c-4576-a6b3-408932ac7bb5.png)
> 2. @ckandag - Should we tweak step 16 a little to say something like "If you are installing into a GCP shared VPC, the VPC name and subnets are being shared from the host project..."
>    ![image](https://user-images.githubusercontent.com/130616558/279811446-47f5f37f-0e65-47a3-83b3-bf011b57e795.png)
> 
> @jianli-wei - can you please review the doc changes and give a QE 👍
> 
> Thanks @mletalie for driving the documentation changes for this feature.



> LGTM.
> 
> A few minor comments -
> 
> 1. Does it makes sense to repeat this note on [Deleting an OpenShift Dedicated cluster](https://66312--docspreview.netlify.app/openshift-dedicated/latest/osd_install_access_delete_cluster/osd-deleting-a-cluster) page as well?
>    ![image](https://user-images.githubusercontent.com/130616558/279811202-67570ae6-8c3c-4576-a6b3-408932ac7bb5.png)
> 2. @ckandag - Should we tweak step 16 a little to say something like "If you are installing into a GCP shared VPC, the VPC name and subnets are being shared from the host project..."
>    ![image](https://user-images.githubusercontent.com/130616558/279811446-47f5f37f-0e65-47a3-83b3-bf011b57e795.png)
> 
> @jianli-wei - can you please review the doc changes and give a QE 👍
> 
> Thanks @mletalie for driving the documentation changes for this feature.

Hello @svmrh,
For point #1, I have added a note to that doc: 
![image](https://github.com/openshift/openshift-docs/assets/122639474/e7c57191-4fb6-45db-965d-a73739f86747)

For point #2, I have added a note: 
![image](https://github.com/openshift/openshift-docs/assets/122639474/97ce29c4-cf71-4bd2-b66c-edb02320bb55)

svmrh commented:
Thanks @mletalie and @jianli-wei

svmrh commented:
**Do not publish these changes just yet**. 
- Engineering team is facing issues provisioning an OSD cluster into GCP Shared VPC with 4.14.x images (no issues seen in case of 4.13.x images). So, until this is resolved, do not publish the content changes for this feature. 
- Slack thread used for triaging this issue: https://redhat-internal.slack.com/archives/CCX9DB894/p1698949140705799

svmrh commented:
[Thursday, Nov 2nd 2:55pm PT] Update: 
The issue has been resolved. After applying the fix, 4.14 OSD cluster install into a GCP Shared VPC was successful. But given the last minute hiccup, **the plan is to go GA on Monday**. 

**We will wait for QE to run sanity tests in production env and confirm if everything looks good**. 

I will update this thread on Monday once everything looks good.

jneczypor commented on modules/deleting-cluster.adoc:
```diff
@@ -19,3 +19,11 @@ You can delete your {product-title} cluster in {cluster-manager-first}.
 . Type the name of the cluster highlighted in bold, then click *Delete*. Cluster deletion occurs automatically.
+
++
+
```
I don't think you need the extra line here.

jneczypor commented on modules/deleting-cluster.adoc:
```diff
@@ -19,3 +19,11 @@ You can delete your {product-title} cluster in {cluster-manager-first}.
+
+[NOTE]
+====
+If you delete a cluster that was installed into a GCP shared VPC, you must inform the VPC owner of the host project to remove the IAM policy roles granted to the service account that was referenced during the creation of the cluster.
```
```suggestion
If you delete a cluster that was installed into a GCP shared VPC, inform the Shared VPC Admin to remove the IAM policy roles granted to the service account that was referenced during cluster creation.
```
We use "VPC Owner" as a proper noun on the most recently updated Shared VPC page in ROSA (https://docs.openshift.com/rosa/rosa_install_access_delete_clusters/rosa-shared-vpc-config.html), but it looks like Google uses "Shared VPC Admin" (https://cloud.google.com/vpc/docs/provisioning-shared-vpc#set-up-shared-vpc), hence the name change.

jneczypor commented on modules/osd-create-cluster-ccs.adoc:
```diff
@@ -204,30 +204,59 @@ The *Use a PrivateLink* option cannot be changed after a cluster is created.
++
+[IMPORTANT]
+====
+
```
Remove the extra line.

jneczypor commented on modules/osd-create-cluster-ccs.adoc:
```diff
@@ -204,30 +204,59 @@ The *Use a PrivateLink* option cannot be changed after a cluster is created.
+[IMPORTANT]
+====
+
+To install a cluster into a shared VPC, you must use {product-title} version 4.13.15 or above. Additionally, the shared VPC administrator must enable a project as a host project in their Google Cloud console. For more information, see link:https://cloud.google.com/vpc/docs/provisioning-shared-vpc#set-up-shared-vpc[Enable a host project].
```
```suggestion
To install a cluster into a shared VPC, you must use {product-title} version 4.13.15 or above. Additionally, the Shared VPC Admin must enable a project as a host project in their Google Cloud console. For more information, see link:https://cloud.google.com/vpc/docs/provisioning-shared-vpc#set-up-shared-vpc[Enable a host project].
```

jneczypor commented on modules/osd-create-cluster-ccs.adoc:
```diff
@@ -204,30 +204,59 @@ The *Use a PrivateLink* option cannot be changed after a cluster is created.
+====
+
+.. Select *Install into GCP shared VPC*.
+.. Specify the **Host project ID**. If the specified host project ID is incorrect, cluster creation will fail.
```
Is "host project ID" initial capped or not? You have it written two different ways here.

EricPonvelle replied:
More on this, since it's bold, I am assuming that it's a field name. Is that correct?  If so, I think the rest is fine, though we may want to mirror how it displays throughout the UI.

We tend to avoid future tense too, so maybe "cluster creation fails."

mletalie replied:
Yes, it is field name. Thanks all! 
![image](https://github.com/openshift/openshift-docs/assets/122639474/471488aa-8b5a-442e-a1bd-22a3b6d97ac3)

jneczypor commented on modules/osd-create-cluster-ccs.adoc:
```diff
@@ -204,30 +204,59 @@ The *Use a PrivateLink* option cannot be changed after a cluster is created.
+.. Select *Install into GCP shared VPC*.
+.. Specify the **Host project ID**. If the specified host project ID is incorrect, cluster creation will fail.
++
+
```
I don't think the extra line is necessary here.

jneczypor commented on modules/osd-create-cluster-ccs.adoc:
```diff
@@ -204,30 +204,59 @@ The *Use a PrivateLink* option cannot be changed after a cluster is created.
+
+[IMPORTANT]
+====
+Once you complete the steps within the cluster configuration wizard and click **Create Cluster**, the cluster will go into the "Installation Waiting" state. At this point, you must contact the VPC administrator of the host project, who must assign the dynamically-generated service account the following roles: **Computer Network Administrator**, **Compute Security Administrator**, and **DNS Admini ...
```
```suggestion
Once you complete the steps within the cluster configuration wizard and click *Create Cluster*, the cluster will go into the "Installation Waiting" state. At this point, you must contact the Shared VPC Admin, who must assign the dynamically-generated service account the following roles: *Computer Network Administrator*, *Compute Security Administrator*, and *DNS Administrator*.
```

jneczypor commented on modules/osd-create-cluster-ccs.adoc:
```diff
@@ -204,30 +204,59 @@ The *Use a PrivateLink* option cannot be changed after a cluster is created.
+[IMPORTANT]
+====
+Once you complete the steps within the cluster configuration wizard and click **Create Cluster**, the cluster will go into the "Installation Waiting" state. At this point, you must contact the VPC administrator of the host project, who must assign the dynamically-generated service account the following roles: **Computer Network Administrator**, **Compute Security Administrator**, and **DNS Admini ...
+The administrator has 30 days to grant the listed permissions before the cluster creation fails.
```
```suggestion
The Shared VPC Admin has 30 days to grant the listed permissions before cluster creation fails.
```

jneczypor commented on modules/osd-create-cluster-ccs.adoc:
```diff
@@ -204,30 +204,59 @@ The *Use a PrivateLink* option cannot be changed after a cluster is created.
+====
+Once you complete the steps within the cluster configuration wizard and click **Create Cluster**, the cluster will go into the "Installation Waiting" state. At this point, you must contact the VPC administrator of the host project, who must assign the dynamically-generated service account the following roles: **Computer Network Administrator**, **Compute Security Administrator**, and **DNS Admini ...
+The administrator has 30 days to grant the listed permissions before the cluster creation fails.
+For information about shared GPC VPC permissions, see link:https://cloud.google.com/vpc/docs/provisioning-shared-vpc#migs-service-accounts[Provision shared VPC].
```
```suggestion
For information about GCP shared VPC permissions, see link:https://cloud.google.com/vpc/docs/provisioning-shared-vpc#migs-service-accounts[Provision Shared VPC].
```
The Google Docs title cap their headings, so I wanted it to be consistent.

jneczypor commented on modules/osd-create-cluster-ccs.adoc:
```diff
@@ -204,30 +204,59 @@ The *Use a PrivateLink* option cannot be changed after a cluster is created.
 GCP
 endif::osd-on-gcp[]
-VPC, provide your *Virtual Private Cloud (VPC) subnet settings* and select *Next*. You must have created the Cloud network address translation (NAT) and a Cloud router. See the additional resources for information about Cloud NATs and Google VPCs.
+VPC, provide your *Virtual Private Cloud (VPC) subnet settings* and select *Next*.
```
```suggestion
VPC, provide your *VPC subnet settings* and select *Next*.
```
Not sure why VPC was spelled out all the way down here. There are other, better "first mentions".

mletalie replied:
![image](https://github.com/openshift/openshift-docs/assets/122639474/01f67332-984b-4476-bc3d-197a2df283d1)
Good point. FWIW, I think it was originally written this way b/c of the way it was presented in the UI.

jneczypor commented on modules/osd-create-cluster-ccs.adoc:
```diff
@@ -284,6 +313,13 @@ In the event of critical security concerns that significantly impact the securit
+ifdef::osd-on-gcp[]
+[NOTE]
+====
+If you delete a cluster that was installed into a GCP shared VPC, you must inform the VPC owner of the host project to remove the IAM policy roles granted to the service account that was referenced during the creation of the cluster.
```
Make this note match the corrections to the same note above.

jneczypor reviewed (commented):
I took a look and added comments for a few changes. LGTM!

EricPonvelle reviewed (commented):
I just had one small follow up, but looks good.

mletalie commented:
/remove-label peer-review-needed

mletalie commented:
/label peer-review-done

svmrh commented:
Both backend and UI changes for this feature are enabled in production. Even the feature flag is removed for all orgs. Please kickoff the process for publishing the content updates for XCMSTRAT-91

EricPonvelle reviewed (approved):
/lgtm

EricPonvelle commented:
/cherrypick enterprise-4.14

EricPonvelle commented:
/cherrypick enterprise-4.15
